packages/gatsby/src/db/node-tracking.js

```diff
@@ -15,30 +15,71 @@ const getRootNodeId = node => rootNodeMap.get(node)
  * @see trackInlineObjectsInRootNode
  * @param {(Object|Array)} data Inline object or array
  * @param {string} nodeId Id of node that contains data passed in first parameter
+ * @param {boolean} sanitize Wether to strip objects of unuspported and not serializable fields
+ * @param {string} [ignore] Fieldname that doesn't need to be tracked and sanitized
+ *
  */
-const addRootNodeToInlineObject = (data, nodeId) => {
-  if (_.isPlainObject(data) || _.isArray(data)) {
-    _.each(data, o => addRootNodeToInlineObject(o, nodeId))
-    rootNodeMap.set(data, nodeId)
+const addRootNodeToInlineObject = (data, nodeId, sanitize, isNode = false) => {
+  const isPlainObject = _.isPlainObject(data)
+
+  if (isPlainObject || _.isArray(data)) {
+    let returnData = data
+    if (sanitize) {
+      returnData = isPlainObject ? {} : []
+    }
+    let anyFieldChanged = false
+    _.each(data, (o, key) => {
+      if (isNode && key === `internal`) {
+        returnData[key] = o
+        return
+      }
+      returnData[key] = addRootNodeToInlineObject(o, nodeId, sanitize)
+
+      if (returnData[key] !== o) {
+        anyFieldChanged = true
+      }
+    })
+
+    if (anyFieldChanged) {
+      if (isPlainObject) {
```

Comment:
```suggestion
      data = omitUndefined(returnData);
```

```js
/**
 * @param {Object} data
 * @param {Object} data without undefined values
 */
function omitUndefined(data) {
  const isPlainObject = _.isPlainObject(data)
  if (isPlainObject) {
    return _.pickBy(returnData, p => p !== undefined)
  }

  return returnData.filter(p => p !== undefined)
}

Reply:
addressed in https://github.com/gatsbyjs/gatsby/pull/10732/commits/1905f2ee5207bf8662fe908cdf8639d89ff9bcbf